test-harness/tests/test_master.py

```diff
@@ -204,3 +205,36 @@ def test_if_upstream_headers_are_correct(self,
                                              valid_user_header,
                                              path_fmt.format(endpoint_type),
                                              )
+
+
+class TestNginxResolver:
+
+    def test_if_mesos_leader_is_resolved(self,
+                                         master_ar_process,
+                                         valid_user_header,
+                                         dns_server_mock,
+                                         ):
+        # Make a request that should be forwarded to Mesos leader instance
+        url = master_ar_process.make_url_from_path('/mesos/master/state-summary')
+
+        # Make sure that request was proxied to MesosEndpoint mock which returns
+        # data that looks like actual cluster state summary
+        r = requests.get(url, allow_redirects=True, headers=valid_user_header)
+        assert r.status_code == 200
+        assert r.json()["cluster"] == "prozlach-qzpz04t"
+
+        # See MockerBase::_create_common_endpoints() where we create ReflectingEndpoint
+        dns_server_mock.reply_with_leader_ip('127.0.0.5')
```

Comment:
We really need this test to be launched in a separate AR instance that is GCed after the tests finishes.

If we do it using shared AR instance (long story short - all that is in the `tests_master.py` file, more info in the README.md) we are risking creating HeisenBugs/flaky tests as change of leader.mesos may be cached i.e. by lua code.